rust-runtime/aws-smithy-runtime/src/client/identity/cache/lazy.rs

```diff
@@ -0,0 +1,671 @@
+        self
+    }
+
+    /// Default expiration time to set on an identity if it doesn't have an expiration time.
+    ///
+    /// This is only used if the resolved identity doesn't have an expiration time set.
+    /// This must be at least 15 minutes.
+    ///
+    /// Defaults to 15 minutes.
+    pub fn default_expiration(mut self, duration: Duration) -> Self {
+        self.set_default_expiration(Some(duration));
+        self
+    }
+
+    /// Default expiration time to set on an identity if it doesn't have an expiration time.
+    ///
+    /// This is only used if the resolved identity doesn't have an expiration time set.
+    /// This must be at least 15 minutes.
+    ///
+    /// Defaults to 15 minutes.
+    pub fn set_default_expiration(&mut self, duration: Option<Duration>) -> &mut Self {
+        self.default_expiration = duration;
+        self
+    }
+
+    /// Builds a [`SharedIdentityCache`] from this builder.
+    ///
+    /// # Panics
+    ///
+    /// This builder will panic if required fields are not given, or if given values are not valid.
+    pub fn build(self) -> SharedIdentityCache {
+        let default_expiration = self.default_expiration.unwrap_or(DEFAULT_EXPIRATION);
+        assert!(
+            default_expiration >= DEFAULT_EXPIRATION,
+            "default_expiration must be at least 15 minutes"
+        );
+        LazyCache::new(
+            self.time_source.unwrap_or_default(),
+            self.sleep_impl.unwrap_or_else(|| {
+                default_async_sleep().expect("no default sleep implementation available")
+            }),
+            self.load_timeout.unwrap_or(DEFAULT_LOAD_TIMEOUT),
+            self.buffer_time.unwrap_or(DEFAULT_BUFFER_TIME),
+            self.buffer_time_jitter_fraction
+                .unwrap_or(DEFAULT_BUFFER_TIME_JITTER_FRACTION),
+            default_expiration,
+        )
+        .into_shared()
+    }
+}
+
+#[derive(Debug)]
+struct CachePartitions {
+    partitions: RwLock<HashMap<IdentityCachePartition, ExpiringCache<Identity, BoxError>>>,
+    buffer_time: Duration,
+}
+
+impl CachePartitions {
+    fn new(buffer_time: Duration) -> Self {
+        Self {
+            partitions: RwLock::new(HashMap::new()),
+            buffer_time,
+        }
+    }
+
+    fn partition(&self, key: IdentityCachePartition) -> ExpiringCache<Identity, BoxError> {
+        let mut partition = self.partitions.read().unwrap().get(&key).cloned();
+        if partition.is_none() {
+            let mut partitions = self.partitions.write().unwrap();
+            // Another thread could have inserted the partition before we acquired the lock,
+            // so double check before inserting it.
+            partitions
+                .entry(key)
+                .or_insert_with(|| ExpiringCache::new(self.buffer_time));
+            drop(partitions);
+
+            partition = self.partitions.read().unwrap().get(&key).cloned();
```

Comment:
since this pattern only works for append-only caches, it may be worth noting that entries are never removed.